### PR TITLE
MediaLibrary: MediaExporter protocol, extension, errors, structs and tests.

### PR DESCRIPTION
**Pull request by kurzee**
Continuing from https://github.com/wordpress-mobile/WordPress-iOS/pull/7116, this PR introduces the base declaration of the `MediaExporter` protocol, it's default extension and supporting declarations.

Note: Exporters are what handle the grunt work of taking whatever Media is selected from whatever source, and processing it for upload. For the most part, I've followed all the hard work already put in to the existing implementation, then refined and consolidated.

- The `MediaExporter` protocol defines what's required of any conforming exporters, such as image exporters, URL exporters, asset exporters, etc.
- The `MediaExporter` `extension` implements a couple default helper methods shared among exporters.
- The `MediaExportError` protocol defines the required conformance for any errors generated by an exporter, which is also handy as a value type.
- `MediaExportSystemError` is a generic `MediaExportError` used for wrapping general system errors encountered by an exporter.
- The `MediaExport` protocol and conforming structs defined what is required of an export that results with any of those particular kind of exports, such as an image or video.

To test:
1. Make sure the code compiles.
2. Run the tests for `MediaExporterTests.swift`

Needs review: @frosty and @SergioEstevao round two :bowtie:

### CODE REVIEW

File: WordPress/Classes/Utility/Media/MediaExporter.swift [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+            return MediaExportSystemError.failedWith(systemError: error)
+        }
+    }
+
+    /// Returns the size of the file at the URL, if available.
+    ///
+    /// - param URL: A file URL.
+    /// - returns: The size in bytes, or nil if unavailable.
+    ///
+    func fileSizeAtURL(_ url: URL) -> Int64? {
```
Comment (SergioEstevao) [inline WordPress/Classes/Utility/Media/MediaExporter.swift]: I feel this method could be an extension to the URL class.

File: WordPress/Classes/Utility/Media/MediaExporter.swift [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+        guard url.isFileURL else {
+            return nil
+        }
+        let attributes = try? FileManager.default.attributesOfItem(atPath: url.path)
```
Comment (SergioEstevao) [inline WordPress/Classes/Utility/Media/MediaExporter.swift]: You can also do url.resourceValues(forKeys:[.fileSizeKey]) to get the file size and avoid the use of the FileManager and paths.

File: WordPress/Classes/Utility/Media/MediaExporter.swift [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+    }
+    func toNSError() -> NSError {
+        switch self {
+        case .failedWith(let systemError):
+            return systemError as NSError
+        }
+    }
+}
+
+protocol MediaExporter {
```
Comment (frosty) [inline WordPress/Classes/Utility/Media/MediaExporter.swift]: Can we add a quick doc comment for the protocol itself?

File: WordPress/Classes/Utility/Media/MediaExporter.swift [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+    ///
+    var maximumImageSize: CGFloat? { get set }
+
+    /// Strip the geoLocation from assets, if needed.
+    ///
+    var stripsGeoLocationIfNeeded: Bool { get set }
+
+    /// The type of MediaDirectory to use for the export destination URL.
+    ///
+    /// - Note: This would almost always be set to .uploads, but for unit testing we use .temporary.almost
```
Comment (frosty) [inline WordPress/Classes/Utility/Media/MediaExporter.swift]: Typo: extra `almost` at the end

File: WordPress/Classes/Utility/Media/MediaExporter.swift (hunk 1 above)
Comment (frosty) [inline WordPress/Classes/Utility/Media/MediaExporter.swift]: Could this just be a `FileManager` extension?

File: WordPress/Classes/Utility/Media/MediaExporter.swift [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        let attributes = try? FileManager.default.attributesOfItem(atPath: url.path)
+        return attributes?[.size] as? Int64
+    }
+
+    /// The expected file extension string for a given UTType identifier.
+    ///
+    /// - param type: The UTType identifier
+    /// - returns: The expected file extension or nil if unknown.
+    ///
+    func fileExtensionForUTType(_ type: String) -> String? {
```
Comment (frosty) [inline WordPress/Classes/Utility/Media/MediaExporter.swift]: Could this just be a `FileManager` extension?

File: WordPress/Classes/Utility/Media/MediaExporter.swift (hunk 2 above)
Comment (kurzee) [inline WordPress/Classes/Utility/Media/MediaExporter.swift]: I think I like that better 👍

**Comment by kurzee** [comment]
@SergioEstevao @frosty after thinking about it for a bit, something feels iffy about moving `fileExtensionForUTType` and `fileSizeAtURL` to a generic extension of `String`, `URL`, or `FileManager`.

Particularly, these methods are slightly tailored towards the MediaExporters and not necessarily intended as a _global_ function to be used or supported otherwise.

I think I'd like to keep these in there for now, but let me know if you think otherwise.

**Comment by kurzee** [comment]
Ready for another look 👾

**Comment by kurzee** [comment]
>I still don't reeeeeally like them in there, but I understand your point.

@frosty I think I've found a way to feel comfortable about it 😄, one more peak?

File: WordPress/Classes/Extensions/URL+Helpers.swift [hunk 6]
```diff
@@ -2,6 +2,18 @@ import Foundation
 
 extension URL {
 
+    /// The URLResource fileSize of the file at the URLs, if available.
+    ///
+    /// - param URL: A file URL.
+    ///
+    var resourceFileSize: Int? {
```
Comment (SergioEstevao) [inline WordPress/Classes/Extensions/URL+Helpers.swift]: fileSize maybe will be a better name. 
And is Int large enough to hold the result? on objective C the fileSize use to be a `long long` type that normally converts to an `Int64`

The parameter documentation also needs to be updated. What is the result of it? bytes? Kbs? and we no longer need the param info.

Comment (kurzee) [inline WordPress/Classes/Extensions/URL+Helpers.swift]: >fileSize maybe will be a better name.

I'd prefer to keep the `resource` keyword as it includes a bit more information as to where this value is/should be coming from.

>And is Int large enough to hold the result?

I did question that, but the `fileSize` attribute does indeed return an `Int`: https://developer.apple.com/reference/foundation/urlresourcevalues/1779734-filesize

>The parameter documentation also needs to be updated. What is the result of it? bytes? Kbs? and we no longer need the param info.

Documentation updated.

**Pull request merged**